rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+  TF_SetStatus(status, TF_OK, "");
+}
+```
+
+## TFRT integration
+
+Current preferred direction would generate a
+[BEF](https://github.com/tensorflow/runtime/blob/master/documents/binary_executable_format.md) file in advance and then run that
+file on a mobile device. Generated BEF file would have to call either native, TF
+Lite, runtime fallback or kernel fallback kernels and provide any glue logic in
+between (such as tensor conversions).
+
+We also need to consider how kernel or runtime fallback will be selected. This
+could be a parameter at BEF file creation step. It might also be good to package
+both runtime and kernel fallback implementations in a BEF file to be selected at
+runtime.
+
+## Size Reduction
+
+Since we want to run on a mobile platform, we need to look for any opportunity
+to cut down size. First of all, we remove dependency on current TensorFlow
+runtime (for e.g. we no longer depend on `NodeDef` and `OpDef` protos). We are
+also looking at ways to reduce large size contributions of
+[absl libraries](https://github.com/abseil/abseil-cpp/tree/master/absl) and
+[protos](https://github.com/protocolbuffers/protobuf).
+
+### Protos
+
+We are currently investigating the following options:
+
+*   Switch to [micropb](https://github.com/protocolbuffers/upb). This proto
+    implementation provides C interfaces and is more compact.
+*   Remove dependency on protos.
+
+### ABSL
+
+We can hide ABSL references behind aliases (see
+[tensorflow::StringPiece](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/stringpiece.h;l=33;drc=af7fd02ca40f362c4ac96dd064d6a2224b65d784)
+for example) to make it easier to replace all references to save binary size.
+
+@gunhan is also starting an effort to define a library of STL utilities that
+helps us cut down on binary size.
+
+## Selecting which kernels to register
+
+We want to add a script to the TF Lite build setup that can determine required
+kernels based on a model. We would then only build these kernels. For now, we
+will only support selective registration when building from source.
+
+Script details still need to be worked out.
+
+### Alternatives Considered
+
+The main alterantive to TFRT Kernel Fallback is TFRT Runtime Fallback. TFRT
+Runtime Fallback will call TensorFlow Eager C API (corresponding RFC should be
+published soon). Main trade offs between the two fallbacks are described in the
+table below:
+
+Property    | TFRT Kernel Fallback  | TFRT Runtime Fallback
+----------- | --------------------- | ---------------------
+Generality  | Support subset of ops | Support all ops
+Performance | Lower overhead        | Higher overhead
+Binary size | Lower (no TF runtime) | Higher
+
+### Performance Implications
+
+*   Slow down due to adding inheritance for `OpKernelContext` and
+    `OpKernelConsturction`.
+*   Speed up for lighter weight kernel calls.
+
+We will run benchmarks to check performance numbers as we work on the
```

Comment:
do we plan to share some numbers (binary size, and performance) as part of the RFC?

Reply:
Other then inside the RFC can we propose to publish these numbers as the output of CI? We are still waiting to find a way to publish regular results with current runtime (see https://github.com/tensorflow/build/issues/10)

Reply:
Good questions. I will try to come up with better way to collect these numbers and share exact details in the next ways.
Sorry for being sloppy about this.

Reply:
I added details for size benchmark in Appendix 3.
To do so, I put opensource prototype of Kernel Fallback here: https://github.com/annarev/tensorflow/tree/kernel_fallback/tensorflow/core/tfrt_fallback/kernel

Based on the opensource prototype, updated size measurements are as follows:

* 900k framework
* 200k per kernel per type

Currently, this is not a propertly setup continuous benchmark, but rather a one-off. Setting up a continuous one would definitely be useful and something we should do once Kernel Fallback is properly committed to TensorFlow.

Performance numbers are harder to compare since this proposal is adding more kernels to run on mobile which we currently don't run on mobile. So, we don't have a good baseline.